### PR TITLE
doc/test-framework/scorecard: update scorecard documentation

### PR DESCRIPTION
**Pull request by AlexNPavel**
**Description of the change:** Update scorecard documentation

**Motivation for the change:** Keep scorecard docs up to date

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@
...
 - Unify CLI debug logging under a global `--verbose` flag ([#1361](https://github.com/operator-framework/operator-sdk/pull/1361))
 - [Go module](https://github.com/golang/go/wiki/Modules) support by default for new Go operators and during Ansible and Helm operator migration. The dependency manager used for a new operator can be explicitly specified for new operators through the `--dep-manager` flag, available in [`operator-sdk new`](https://github.com/operator-framework/operator-sdk/blob/master/doc/sdk-cli-reference.md#new) and [`operator-sdk migrate`](https://github.com/operator-framework/operator-sdk/blob/master/doc/sdk-cli-reference.md#migrate). `dep` is still available through `--dep-manager=dep`. ([#1001](https://github.com/operator-framework/operator-sdk/pull/1001))
+- Operator Scorecard plugin support ([#1379](https://github.com/operator-framework/operator-sdk/pull/1379)). Documention for scorecard plugins can be found in the main scorecard doc: [doc/test-framework/scorecard.md](./doc/test-framework/scorecard.md)
```
Comment (hasbro17) [inline CHANGELOG.md]:
```suggestion
- Operator Scorecard plugin support ([#1379](https://github.com/operator-framework/operator-sdk/pull/1379)). Documentation for scorecard plugins can be found in the main scorecard doc: [doc/test-framework/scorecard.md](./doc/test-framework/scorecard.md)
```

File: doc/sdk-cli-reference.md [hunk 2]
```diff
@@ -437,34 +439,28 @@ Run scorecard tests on an operator
 * `namespaced-manifest` string - Path to manifest for namespaced resources (e.g. RBAC and Operator manifest)
 * `olm-deployed` - Only use the CSV at `csv-path` for manifest data, except for those provided to `cr-manifest`
 * `olm-tests` - Enable OLM integration checks (default true)
+* `-o, --output` string - Output format for results. Valid values: human-readable, json (default "human-readable")
```
Comment (hasbro17) [inline doc/sdk-cli-reference.md]: Nit: Just making the keywords more clear.
```suggestion
* `-o, --output` string - Output format for results. Valid values: `human-readable` or `json` (default `human-readable`)
```

File: doc/test-framework/scorecard.md [hunk 3]
```diff
@@ -8,8 +8,10 @@ stage of development, it will gain more functionality and stabilize over time.
 
 The scorecard works by creating all resources required by CRs and the operator. For the operator
 deployment, it also adds another container to the operator's pod that is used to record calls to the API server,
-which are analyzed by the scorecard for various tests. The scorecard will also analyze the CR object itself,
-modifying spec fields and monitoring how the operator responds.
+which are analyzed by the scorecard for various tests. The scorecard will also look at some of the fields in the
+CR object itself for some of the tests.
+
+As of `v0.8.0` scorecard also supports plugins. This allows external developers extend the functionality of the scorecard and add additional tests.
```
Comment (hasbro17) [inline doc/test-framework/scorecard.md]:
```suggestion
As of `v0.8.0` scorecard also supports plugins. This allows external developers to extend the functionality of the scorecard and add additional tests.
```

File: doc/test-framework/scorecard.md [hunk 4]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
...
-This test makes sure that the CRDs listed in the [`owned` CRDs section][owned-crds] of the CSV have a `resources` subsection. If the
+This test makes sure that the CRDs for each CR provided via the `--cr-manifest` flag have a `resources` subsection in the [`owned` CRDs section][owned-crds] of the CSV. If the
 test detects used resources that are not listed in the resources section, it will list them in the suggestions at the end of the test.
-This test has a maximum score equal to the number of CRDs listed in the CSV.
+This test has a maximum score equal to the number of CRs provided via the `--cr-manifest` flag.
 
 #### CRs Have At Least 1 Example
 
-This test checks that the CSV has an [`alm-examples` section][alm-examples] in its metadatas' annotations. This test has a maximum score of 1.
+This test checks that the CSV has an [`alm-examples` section][alm-examples] for each CR provided by the `--cr-manifest` flag in its metadatas' annotations. This test has a maximum score
```
Comment (hasbro17) [inline doc/test-framework/scorecard.md]:
```suggestion
This test checks that the CSV has an [`alm-examples` section][alm-examples] for each CR provided by the `--cr-manifest` flag in its metadata's annotations. This test has a maximum score
```

File: doc/test-framework/scorecard.md [hunk 5]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
...
+For instance, `scorecard/assets`. The scorecard runs all plugins from the root `scorecard` directory. Since the scorecard will run all executable files in the scorecard's
+`bin` subdirectory, the plugins can be written in any programming language supported by the OS the scorecard is being run on.
+
+To provide results to the scorecard, the plugin must output a valid JSON object to its `stdout`. Invalid JSON in the `stdout` will result in the plugin being marked as failed.
+To provide logs to the scorecard, plugins can either set the `log` field for the scorecard suites they return or they can output logs to `stderr`, which will stream the log
+to the console if the scorecard is being run in with `human-readable` output or be added to the main ScorecardOutput `log` field when being run with the output flag set to `json`.
+
+### JSON format
+
+The JSON output is formatted in the same way that a Kubernetes API would be, which allows us to make updates and new versions of the schema if necessary as well as use various
```
Comment (hasbro17) [inline doc/test-framework/scorecard.md]: Trying to avoid the use of "us".
```suggestion
The JSON output is formatted in the same way that a Kubernetes API would be, which allows for updates to the schema as well as the use of various
```

File: doc/test-framework/scorecard.md [hunk 6]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
...
+        }
+      },
+      "additionalProperties": false,
+      "type": "object"
+    }
+  }
+}
+```
+
+NOTE: The `ScorecardOutput` object is designed the same as a kubernetes API, and thus also has a full `TypeMeta` and `ObjectMeta`. Those are optional and not
```
Comment (hasbro17) [inline doc/test-framework/scorecard.md]: Nit: Highlight the note with **bold**.
```suggestion
**NOTE:** The `ScorecardOutput` object is designed the same as a Kubernetes API, and thus also has a full `TypeMeta` and `ObjectMeta`. Those are optional and not
```

File: doc/test-framework/scorecard.md [hunk 7]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
...
+          "suggestions": [],
+          "errors": []
+        }
+      ]
+    }
+  ]
+}
+```
+
+NOTE: The `ScorecardOutput.Log` field is only intended to be used to log the scorecard's output and the scorecard will ignore that field if a plugin provides it.
```
Comment (hasbro17) [inline doc/test-framework/scorecard.md]:
```suggestion
**NOTE:** The `ScorecardOutput.Log` field is only intended to be used to log the scorecard's output and the scorecard will ignore that field if a plugin provides it.
```

File: doc/test-framework/scorecard.md [hunk 8]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
...
+`bin` subdirectory, the plugins can be written in any programming language supported by the OS the scorecard is being run on.
+
+To provide results to the scorecard, the plugin must output a valid JSON object to its `stdout`. Invalid JSON in the `stdout` will result in the plugin being marked as failed.
+To provide logs to the scorecard, plugins can either set the `log` field for the scorecard suites they return or they can output logs to `stderr`, which will stream the log
+to the console if the scorecard is being run in with `human-readable` output or be added to the main ScorecardOutput `log` field when being run with the output flag set to `json`.
+
+### JSON format
+
+The JSON output is formatted in the same way that a Kubernetes API would be, which allows us to make updates and new versions of the schema if necessary as well as use various
+kubernetes helpers. The Golang structs are defined in `pkg/apis/scorecard/v1alpha1/types.go` and can be easily implemented by plugins written in Golang. Below is the JSON Schema:
```
Comment (hasbro17) [inline doc/test-framework/scorecard.md]:
```suggestion
Kubernetes helpers. The Golang structs are defined in `pkg/apis/scorecard/v1alpha1/types.go` and can be easily implemented by plugins written in Golang. Below is the JSON Schema:
```

File: doc/sdk-cli-reference.md [hunk 9]
```diff
@@ -428,7 +428,9 @@ Run scorecard tests on an operator
 ### Flags
 
 * `basic-tests` - Enable basic operator checks (default true)
+* `config` string - config file (default is `<project_dir>/.osdk-yaml`)
```
Comment (joelanford) [inline doc/sdk-cli-reference.md]: Should the default filename be `<project_dir>/.osdk-scorecard`?

https://github.com/operator-framework/operator-sdk/blob/d4d37b10fc4bd875f56d75d59c69b3e53674a304/internal/pkg/scorecard/scorecard.go#L455

File: doc/test-framework/scorecard.md [hunk 10]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
...
+      "additionalProperties": false,
+      "type": "object"
+    }
+  }
+}
+```
+
+**NOTE:** The `ScorecardOutput` object is designed the same as a Kubernetes API, and thus also has a full `TypeMeta` and `ObjectMeta`. Those are optional and not
+used by the scorecard at the current time. These optional fields have thus been removed from the schema for cleanliness and the schema only includes the fields
+the scorecard currently looks at.
```
Comment (joelanford) [inline doc/test-framework/scorecard.md]: Do you think we should make these required now and start checking for them? The other option is to allow them to be optional for `v1alpha1` and required for all future revisions. I'm okay with either option, but if we go with the latter options, I think we should add another sentence to this note that says these fields will be required in future revisions to this API.

File: doc/sdk-cli-reference.md (hunk 9 above)
Comment (joelanford) [reply doc/sdk-cli-reference.md]: Also, by the way, it looks like this should be fixed here as well:

https://github.com/operator-framework/operator-sdk/blob/d4d37b10fc4bd875f56d75d59c69b3e53674a304/cmd/operator-sdk/scorecard/cmd.go#L39

Comment (AlexNPavel) [reply doc/sdk-cli-reference.md]: Yeah. I think I probably forgot to change the help text when we decided on a default name. I'll fix that.

File: doc/test-framework/scorecard.md (hunk 10 above)
Comment (AlexNPavel) [reply doc/test-framework/scorecard.md]: While we don't actually check `apiVersion` yet (which we should), I meant that field to be required as well as the `kind` (which is why those are marked as required in the JSONSchema). I was referring to other fields in `ObjectMeta`, such as `selfLink`, `uid`, `creationTimestamp`, etc. are not required at the current time. I'll rephrase this to make that clear.

File: doc/test-framework/scorecard.md [hunk 11]
```diff
@@ -54,20 +59,20 @@ The hierarchy of config methods from highest priority to least is: flag->file->d
...
 
-There are 8 tests the scorecard can run:
+There are 8 builtin tests the scorecard can run. If multiple CRs are specified, the test environment is fully cleaned up after each CR so each CR gets a clean testing environment.
 
 ### Basic Operator
 
 #### Spec Block Exists
 
-This test checks the Custom Resource that is created in the cluster to make sure that it has a spec block. This test
+This test checks the Custom Resource(s) that is created in the cluster to make sure that all CRs have a spec block. This test
```
Comment (estroz) [inline doc/test-framework/scorecard.md]:
```suggestion
This test checks the Custom Resource(s) created in the cluster to make sure that all CRs have a spec block. This test
```

File: doc/test-framework/scorecard.md [hunk 12]
```diff
@@ -54,20 +59,20 @@ The hierarchy of config methods from highest priority to least is: flag->file->d
...
 has a maximum score of 1.
 
 #### Status Block Exists
 
-This test checks the Custom Resource that is created in the cluster to make sure that it has a status block. This
+This test checks the Custom Resource(s) that is created in the cluster to make sure that all CRs have a status block. This
```
Comment (estroz) [inline doc/test-framework/scorecard.md]:
```suggestion
This test checks the Custom Resource(s) created in the cluster to make sure that all CRs have a status block. This
```

File: doc/test-framework/scorecard.md [hunk 13]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
 
 #### Provided APIs have validation
 
-This test verifies that all the CRDs in the CRDs folder contain a validation section. If the CRD matches the kind and version of the
-CR currently being tested, it will also verify that there is a validation for each spec and status field in that CR. This test has a
-maximum score of 1.
+This test verifies that the CRDs for the provided CRs contain a validation section and that there is a validation for each
```
Comment (estroz) [inline doc/test-framework/scorecard.md]:
```suggestion
This test verifies that the CRDs for the provided CRs contain a validation section and that there is validation for each
```

The way validation in relation to CRDs is referred to is a bit wonky in general, but this reads the best IMO.

File: doc/test-framework/scorecard.md [hunk 14]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
...
-This test makes sure that the CRDs listed in the [`owned` CRDs section][owned-crds] of the CSV have a `resources` subsection. If the
+This test makes sure that the CRDs for each CR provided via the `--cr-manifest` flag have a `resources` subsection in the [`owned` CRDs section][owned-crds] of the CSV. If the
 test detects used resources that are not listed in the resources section, it will list them in the suggestions at the end of the test.
-This test has a maximum score equal to the number of CRDs listed in the CSV.
+This test has a maximum score equal to the number of CRs provided via the `--cr-manifest` flag.
 
 #### CRs Have At Least 1 Example
 
-This test checks that the CSV has an [`alm-examples` section][alm-examples] in its metadatas' annotations. This test has a maximum score of 1.
+This test checks that the CSV has an [`alm-examples` section][alm-examples] for each CR provided by the `--cr-manifest` flag in its metadata's annotations. This test has a maximum score
```
Comment (estroz) [inline doc/test-framework/scorecard.md]:
```suggestion
This test checks that the CSV has an [`alm-examples` annotation][alm-examples] for each CR passed to the `--cr-manifest` flag in its metadata. This test has a maximum score
```

File: doc/test-framework/scorecard.md [hunk 15]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
...
-This test verifies that every field in the Custom Resource's status section has a corresponding descriptor listed in
-the CSV. This test has a maximum score equal to the number of fields in the status section of your Custom Resource.
+This test verifies that every field in the Custom Resources' status sections have a corresponding descriptor listed in
+the CSV. This test has a maximum score equal to the total number of fields in the status sections of each custom resource passed in via the `--cr-manifest` flag.
+
+## Extending the Scorecard with Plugins
+
+To allow the scorecard to be further extended and capable of more complex testing as well as allow the community to make their own scorecard tests, a plugin system has been implemented
+for the scorecard. To use it, a user simply needs to add a binary or script to a `scorecard/bin` directory in their operator's root directory that runs the plugin. It is
+recommended to run plugins with a script to allow more configuration from the user and place other required assets in another subdirectory in the `scorecard` directory.
```
Comment (estroz) [inline doc/test-framework/scorecard.md]:
```suggestion
recommended to run plugins with a script to allow more configuration from the user and place other required assets in another subdirectory in the `scorecard` directory, ex. `scorecard/assets`.
```

File: doc/test-framework/scorecard.md [hunk 16]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
...
+For instance, `scorecard/assets`. The scorecard runs all plugins from the root `scorecard` directory. Since the scorecard will run all executable files in the scorecard's
```
Comment (estroz) [inline doc/test-framework/scorecard.md]:
```suggestion
The scorecard runs all plugins from the root `scorecard` directory. Since the scorecard will run all executable files in the scorecard's
```

File: doc/test-framework/scorecard.md [hunk 17]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
...
+`bin` subdirectory, the plugins can be written in any programming language supported by the OS the scorecard is being run on.
+
+To provide results to the scorecard, the plugin must output a valid JSON object to its `stdout`. Invalid JSON in the `stdout` will result in the plugin being marked as failed.
```
Comment (estroz) [inline doc/test-framework/scorecard.md]:
```suggestion
To provide results to the scorecard, the plugin must output a valid JSON object to its `stdout`. Invalid JSON in `stdout` will result in the plugin being marked as failed.
```

File: doc/test-framework/scorecard.md [hunk 18]
```diff
@@ -79,29 +84,245 @@ API server, indicating that it is modifying resources. This test has a maximum s
...
+To provide logs to the scorecard, plugins can either set the `log` field for the scorecard suites they return or they can output logs to `stderr`, which will stream the log
+to the console if the scorecard is being run in with `human-readable` output or be added to the main ScorecardOutput `log` field when being run with the output flag set to `json`.
```
Comment (estroz) [inline doc/test-framework/scorecard.md]:
```suggestion
to the console if the scorecard is being run in with `--output=human-readable` or be added to the main ScorecardOutput `log` field when being run with `--output=json`.
```

File: doc/sdk-cli-reference.md (hunk 9 above)
Comment (AlexNPavel) [reply doc/sdk-cli-reference.md]: Updated config file usage text

File: doc/test-framework/scorecard.md (hunk 10 above)
Comment (AlexNPavel) [reply doc/test-framework/scorecard.md]: Updated note
